File: creational/abstract-factory.cpp
#include <iostream>

// Abstract Product: Button
class Button
{
public:
  virtual void render() = 0;
  virtual ~Button() {}
};

// Concrete Product: Web Button
class WebButton : public Button
{
public:
  void render() override
  {
    std::cout << "Rendering a web button." << std::endl;
  }
};

// Concrete Product: Mobile Button
class MobileButton : public Button
{
public:
  void render() override
  {
    std::cout << "Rendering a mobile button." << std::endl;
  }
};

// Abstract Product: InputField
class InputField
{
public:
  virtual void render() = 0;
  virtual ~InputField() {}
};

// Concrete Product: Web InputField
class WebInputField : public InputField
{
public:
  void render() override
  {
    std::cout << "Rendering a web input field." << std::endl;
  }
};

// Concrete Product: Mobile InputField
class MobileInputField : public InputField
{
public:
  void render() override
  {
    std::cout << "Rendering a mobile input field." << std::endl;
  }
};

// Abstract Factory Interface
class UIComponentFactory
{
public:
  virtual Button *createButton() = 0;
  virtual InputField *createInputField() = 0;
  virtual ~UIComponentFactory() {}
};

// Concrete Factory for Web
class WebUIComponentFactory : public UIComponentFactory
{
public:
  Button *createButton() override
  {
    return new WebButton();
  }

  InputField *createInputField() override
  {
    return new WebInputField();
  }
};

// Concrete Factory for Mobile
class MobileUIComponentFactory : public UIComponentFactory
{
public:
  Button *createButton() override
  {
    return new MobileButton();
  }

  InputField *createInputField() override
  {
    return new MobileInputField();
  }
};

// Client
class UIManager
{
private:
  UIComponentFactory *_factory;

public:
  UIManager(UIComponentFactory *factory) : _factory(factory) {}

  void renderUI()
  {
    Button *button = _factory->createButton();
    InputField *inputField = _factory->createInputField();

    button->render();
    inputField->render();

    delete button;
    delete inputField;
  }

  ~UIManager()
  {
    delete _factory;
  }
};

// Example Usage
int main()
{
  UIManager *webUIManager = new UIManager(new WebUIComponentFactory());
  webUIManager->renderUI();

  UIManager *mobileUIManager = new UIManager(new MobileUIComponentFactory());
  mobileUIManager->renderUI();

  delete webUIManager;
  delete mobileUIManager;

  return 0;
}

File: creational/abstract-factory.py
from abc import ABC, abstractmethod


# Abstract Factory Interface
class UIComponentFactory(ABC):
    @abstractmethod
    def create_button(self) -> "Button":
        pass

    @abstractmethod
    def create_input_field(self) -> "InputField":
        pass


# Concrete Factory for Web
class WebUIComponentFactory(UIComponentFactory):
    def create_button(self) -> "Button":
        return WebButton()

    def create_input_field(self) -> "InputField":
        return WebInputField()


# Concrete Factory for Mobile
class MobileUIComponentFactory(UIComponentFactory):
    def create_button(self) -> "Button":
        return MobileButton()

    def create_input_field(self) -> "InputField":
        return MobileInputField()


# Abstract Product: Button
class Button(ABC):
    @abstractmethod
    def render(self) -> None:
        pass


# Concrete Product: Web Button
class WebButton(Button):
    def render(self) -> None:
        print("Rendering a web button.")


# Concrete Product: Mobile Button
class MobileButton(Button):
    def render(self) -> None:
        print("Rendering a mobile button.")


# Abstract Product: InputField
class InputField(ABC):
    @abstractmethod
    def render(self) -> None:
        pass


# Concrete Product: Web InputField
class WebInputField(InputField):
    def render(self) -> None:
        print("Rendering a web input field.")


# Concrete Product: Mobile InputField
class MobileInputField(InputField):
    def render(self) -> None:
        print("Rendering a mobile input field.")


# Client
class UIManager:
    def __init__(self, factory: UIComponentFactory) -> None:
        self._factory = factory

    def render_ui(self) -> None:
        button = self._factory.create_button()
        input_field = self._factory.create_input_field()

        button.render()
        input_field.render()


# Example Usage
web_ui_manager = UIManager(WebUIComponentFactory())
web_ui_manager.render_ui()

mobile_ui_manager = UIManager(MobileUIComponentFactory())
mobile_ui_manager.render_ui()
